src/remote_write/src/pooled_parser.rs

```diff
@@ -0,0 +1,221 @@
+//!     }
+//! }
+//! // Object automatically returned to the pool when dropped.
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Note: if you use `decode_async()`, you need to add `deadpool` to your
+//! dependencies:
+//!
+//! ```toml
+//! [dependencies]
+//! deadpool = { workspace = true }  # or "0.10" if not inside this workspace.
+//! ```
+//!
+//! ## Enable Unsafe Optimization
+//!
+//! ```toml
+//! [dependencies]
+//! remote_write = { path = ".", features = ["unsafe-split"] }
+//! ```
+//!
+//! This feature enables zero-copy bytes split that bypasses Rust's memory
+//! safety guarantees. The parsed `Bytes` instances **cannot outlive** the
+//! input raw protobuf `Bytes` instance.
+//!
+//! Correct usage example:
+//!
+//! ```rust,no_run
+//! use bytes::Bytes;
+//! use remote_write::pooled_parser::PooledParser;
+//! # use std::error::Error;
+//! // CORRECT: consume immediately.
+//! async fn foo() -> Result<Vec<String>, Box<dyn Error>> {
+//!     # fn load_protobuf_data() -> Bytes { Bytes::new() }
+//!     let raw_data = load_protobuf_data(); // Bytes.
+//!     let parser = PooledParser::default();
+//!     let request = parser.decode_async(raw_data).await?;
+//!
+//!     let metric_names: Vec<String> = request
+//!         .timeseries
+//!         .iter()
+//!         .flat_map(|ts| ts.labels.iter())
+//!         .filter(|label| label.name.as_ref() == b"__name__")
+//!         .filter_map(|label| String::from_utf8(label.value.to_vec()).ok())
+//!         .collect();
+//!
+//!     Ok(metric_names)
+//! }
+//! ```
+//!
+//! Wrong usage example:
+//!
+//! ```rust,ignore
+//! // WRONG: returned parsed data outlives the input.
+//! fn bar() -> PooledWriteRequest {
+//!     let raw_data = load_protobuf_data(); // Bytes
+//!     let parser = PooledParser::default();
+//!     parser.decode(raw_data).unwrap()
+//!     // The input `raw_data` is dropped here, but the returned `PooledWriteRequest` outlives it.
+//! }
+//! ```
+
+use bytes::Bytes;
+
+use crate::{
+    pb_reader::read_write_request,
+    pooled_types::{PooledWriteRequest, WriteRequestManager, POOL},
+    repeated_field::Clear,
+};
+
+#[derive(Debug, Clone)]
+pub struct PooledParser;
+
+impl PooledParser {
+    fn new() -> Self {
+        Self
+    }
+
+    /// Decode a [`PooledWriteRequest`] from the buffer and return it.
+    ///
+    /// This method will allocate a new [`PooledWriteRequest`] instance since it
+    /// is unable to use the object pool in sync functions.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use bytes::Bytes;
+    /// use remote_write::pooled_parser::PooledParser;
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// # let data = Bytes::new();
+    ///
+    /// let parser = PooledParser::default();
+    /// let request = parser.decode(data)?;
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn decode(&self, buf: Bytes) -> Result<PooledWriteRequest, String> {
```

Comment:
I prefer we use anyhow as the Error type for this crate, it has more features like backtrace.